### PR TITLE
Stable `MonoItem`

### PR DESCRIPTION
**Pull request by ericmarkmartin**

### CODE REVIEW

File: compiler/stable_mir/src/ty.rs [hunk 1]
```diff
@@ -565,3 +565,24 @@ pub enum ImplPolarity {
...
+    VTableShim(DefId),
+    ReifyShim(DefId),
+    FnPtrShim(DefId, Ty),
+    Virtual(DefId, usize),
+    ClosureOnceShim { call_once: DefId, track_caller: bool },
+    ThreadLocalShim(DefId),
+    DropGlue(DefId, Option<Ty>),
+    CloneShim(DefId, Ty),
+    FnPtrAddrShim(DefId, Ty),
+}
```
Comment (bjorn3) [inline compiler/stable_mir/src/ty.rs:588]: I don't think this should expose all the kinds of shims rustc has. This list changes occasionally. Instead I think you should have `Item`, `Intrinsic`, `Virtual` and a generic `Shim` variant which only exposes just enough to get the Stable MIR for the specific instantiation of said shim.

File: compiler/stable_mir/src/ty.rs [hunk 2]
```diff
@@ -565,3 +565,24 @@ pub enum ImplPolarity {
     Negative,
     Reservation,
 }
+
+#[derive(Clone, Debug)]
+pub struct Instance {
+    pub def: InstanceDef,
+    pub args: GenericArgs,
```
Comment (bjorn3) [inline compiler/stable_mir/src/ty.rs:572]: Maybe merge this into the `InstanceDef` variants? For shims this duplicates information with the `Ty` and having it get out of sync could potentially crash the compiler and if it doesn't crash would likely misbehave.

**Comment by oli-obk** [comment]
Can we avoid having MonoItem and `Instance` entirely?

cc @ouz-a this is the same stuff that I've been talking about in https://rust-lang.zulipchat.com/#narrow/stream/320896-project-stable-mir/topic/Normalization.20of.20generic.20ZST/near/395236860

Only that in your case we can avoid it, and I'm hoping we can avoid it here, too, by keeping the result of `Instance::resolve` within the `rustc_smir` crate and only using it to directly invoke `instance_mir` or similar functions.

@ericmarkmartin I'm assuming this is for your experiments in using SMIR in Kani? How does that logic there use `Instance`? Can we just use an `Opaque` in the same place for uniquely identifying an instance, but without any extra usable information in that identifier? It would mean you'd have to store a `Body` immediately, but that may be ok and sufficient

**Pull request closed without merging**